sdk/eventhub/Azure.Messaging.EventHubs/MIGRATIONGUIDE.md

```diff
@@ -0,0 +1,395 @@
+
+## Table of contents
+
+- [Table of contents](#table-of-contents)
+- [General changes](#general-changes)
+  - [Client constructors](#client-constructors)
+  - [Publish events](#publish-events)
+  - [Read events](#read-events)
+- [Migration samples](#migration-samples)
+  - [Migrating code from `PartitionSender` to `EventHubProducerClient` for publishing events to a partition](#migrating-code-from-partitionsender-to-eventhubproducerclient-for-publishing-events-to-a-partition)
+  - [Migrating code from `EventHubClient` to `EventHubProducerAsyncClient` for publishing events using automatic routing](#migrating-code-from-eventhubclient-to-eventhubproducerclient-for-publishing-events-using-automatic-routing)
+  - [Migrating code from `EventHubClient` to `EventHubProducerAsyncClient` for publishing events with partition key](#migrating-code-from-eventhubclient-to-eventhubproducerclient-for-publishing-events-with-partition-key)
+  - [Migrating code from `PartitionReceiver` to `EventHubConsumerAsyncClient` for reading events in batches](#migrating-code-from-partitionreceiver-to-eventhubconsumerclient-for-reading-events-in-batches)
+  - [Migrating code from `EventProcessorHost` to `EventProcessorClient` for reading events](#migrating-code-from-eventprocessorhost-to-eventprocessorclient-for-reading-events)
+- [Additional samples](#additional-samples)
+
+## General changes
+
+In the interest of simplifying the API surface we've made two distinct
+clients, rather than having a single `EventHubClient`:
+* [EventHubProducerClient](https://docs.microsoft.com/en-us/dotnet/api/azure.messaging.eventhubs.eventhubproducerclient?view=azure-dotnet-preview)
+  for publishing messages.
+* [EventHubConsumerClient](https://docs.microsoft.com/en-us/dotnet/api/azure.messaging.eventhubs.eventhubconsumerclient?view=azure-dotnet-preview) 
+  for reading messages.
+
+The producer and consumer clients operate in the context of a specific event hub and offer operations for all partitions. Unlike the v4, the clients are not bound to a specific partition, but the methods on them have overloads to handle specific partitions if needed.
+
+We've introduced a new library [Azure.Messaging.EventHubs.Processor](https://www.nuget.org/packages/Azure.Messaging.EventHubs.Processor/) which replaces the older [Microsoft.Azure.EventHubs.Processor](https://www.nuget.org/packages/Microsoft.Azure.EventHubs.Processor/). Here we have the [EventProcessorClient](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs) which is responsible for consuming events for the configured Event Hub and consumer group across all partitions. It also supports checkpointing and load balancing. Currently, only Azure Storage Blobs is supported for checkpointing.
```

Comment:
>EventProcessorClient [](start = 289, length = 20)

Maybe better to provide link to API doc, not the source.

Reply:
Docs are not properly updated yet and `EventProcessorClient` is missing from the [API doc](https://docs.microsoft.com/en-us/dotnet/api/azure.messaging.eventhubs.processor?view=azure-dotnet-preview&viewFallbackFrom=azure-dotnet), that's why I added the source code link here.

@scbedd could you please take a look into that?

cc: @jsquire, @ramya-rao-a

Reply:
I like the idea.   There's a problem currently with the preview docs for this release where they don't contain the processor package.   We could, however, link out to Dozie's github site in the meantime.

Reply:
Could you please provide the link to the Dozie's github site?

Reply:
I'd be happy to but, uh, I can't remember what it was nor find it.     Do you mind asking him?

Reply:
I think Dozie is OOF. For now we keep as is and I'll update this link later.